[1단계 - 상품 관리 기능] 필립(양재필) 미션 제출합니다.

pilyang opened the pull request:
안녕하세요 오즈!
1단계 미션 제출합니다.

미션을 진행하며 Product의 정보를 다루는데 있어 id값이 필요한 곳이 있고, 없는곳이 있었습니다.

`productService`와 `controller` 사이의 데이터 전송에 상품 데이터를 담기 위해 `ProductDto`를 만들어 사용했습니다. 그런데 경우에 따라 product의 id값이 필요한 경우가 있고, 필요가 없는 경우가 있어 이에대한 dto를 분리해야하나, id가 필요 없는 경우 id값을 null로 놔둬야 하나에 대한 고민을 했습니다.  
페어와 이야기를 해보고, id를 제외한 부분에서 의미적으로도, 코드적으로도 중복이 더 크다고 생각을 하여 지금처럼 빌더로 id값을 필요시에만 입력하도록 하였습니다. 하지만 아직 이게 좋은 선택이였을지 다른 방법은 없었을지에 대해서는 확신이 없어 고민이 있습니다.

추가로 각 계층간에 상품 정보를 전달할때 어떤 객체를 이용해야 하는지에 대한 고민도 많았습니다. 

해당 부분들에서 잘못 생각하고 잘못 만든 부분들이 있으면 따끔하게 지적 부탁드립니다. 다른 부분들 에서도 아직 많이 부족하여 많이 지적해주시면 찾아보고 공부해볼 수 있도록 하겠습니다.

잘부탁드립니다!
감사합니다!!

pilyang committed:
docs: 기능목록 작성

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: db 초기화 설정

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: products crud 구현

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: web controller 연결

- web template 연결
- product 서비스 구현

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: 관리자 상품 정보 관리 기능 구현

- 상품 관리 api구현
 - 상품 추가 구현
 - 상품 수정 구현
 - 상품 제거 구현

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: admin api url 변경

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: dto 생성 전략, 패키지 변경

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: api 컨벤션에 따름

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: request body 검증 추가

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
feat: 상품 업데이트 예외사항 추가

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: 에러 반환 body 변경

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
test: rest api 테스트 코드 추가

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
fix: view - product delete url 수정

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: 상품 관리 api url 수정

Co-authored-by: waterricecake <fruturum@nate.com>

pilyang committed:
refactor: Class 네이밍 변경

- WebController -> ViewController
- AdminControllerAdvice -> ProductControllerAdvice

Co-authored-by: waterricecake <fruturum@nate.com>

ohjoohyung commented on build.gradle:
```diff
@@ -14,6 +14,7 @@ dependencies {
     implementation 'org.springframework.boot:spring-boot-starter-web'
     implementation 'org.springframework.boot:spring-boot-starter-jdbc'
     implementation 'org.springframework.boot:spring-boot-starter-thymeleaf'
+    implementation 'org.springframework.boot:spring-boot-starter-validation'
```
spring validation을 사용해보셨군요 👍
사용하면서 느낀 점 한번 공유해주세요!

pilyang replied:
RuqestBody 에 담겨있는 부분에 대한 검증 코드를 직접 작성하지 않아도 된다는 점에서 편하게 검증을 할 수 있었습니다.

추가로 처음에 해당 spring validation을 적용하기 전에는 request body에 필요한 값이 누락되어 전달되어 npe가 발생한다던가 하면 500 internal server error가 터지지만 `@Valid`를 이용하면 자동으로 bad request 를 반환하고, 필요에 따라 `MethodArguemntNotValidException`으로 쉽게 확인이 되어서 편했습니다.

ohjoohyung replied:
공유해주셔서 감사합니다.
유효성 검증과 관련해서 spring validation을 이용해 dto에서 해줄지, 입력 값으로 도메인을 만들게 될 때 해줄지, 둘 다 해줄지 이런 부분에 대해 계속해서 고민하시게 될 것 같아요.
정답은 없지만 사용하면서 느낀 점을 통해 적절한 기준을 찾아보셨으면 싶어 말씀드려봤습니다..ㅎㅎ

ohjoohyung commented on src/main/java/cart/controller/ProductControllerAdvice.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice(assignableTypes = ProductsApiController.class)
```
ControllerAdvice와 RestControllerAdvice 차이를 한번 공부해보면 좋을 것 같아요.

pilyang replied:
둘은 비교해보니 `@RestControllerAdvice`에는 `@ResponseBody`가 추가되어 있음을 확인했습니다. `@Controller`와 `@RestController`와 같은 차이인데 Controller Advice도 Controller와 마찬가지로 RestControllerAdvice가 있다는 사실을 몰랐고, 생각을 하지 못하였습니다...

해당 advice에서는 오류시 따로 다른 페이지를 보내주는것이 아닌 모든 예외 헨들러가 예외 메시지를 response body에 담아서 보내주는 역할을 하고 있어서 `@RestControllerAdvice`가 더 적합하다고 생각이 들어 해당 어노태이션을 사용하도록 수정했습니다.

만약에 예외시 다른 페이지를 보내주는 역할을 한다면 그때 `@ControllerAdvice`를 사용하는것이 좋을 것 같다고 생각됩니다.

ohjoohyung commented on src/main/java/cart/controller/ProductsApiController.java:
```diff
@@ -0,0 +1,82 @@
+    @PostMapping
+    public void insertProduct(
+            @Valid @RequestBody final ProductRequestDto productRequestDto,
+            final HttpServletResponse httpServletResponse
```
CotnrollerAdvice에서는 ResponseEntity를 사용해보셨더라고요.
여기선 HttpServletResponse를 사용해서 상태 코드를 보내주도록 한 이유를 알 수 있을까요?

pilyang replied:
우선 위에서 ResponseEntity를 사용한 이유는 해당 코멘트에 남긴대로 `@RestControllerAdvice`에 대해서 생각을 해보지 못하여서  사용을 했었습니다.

여기서 ResponseEntity를 사용하지 않은 이유는 페어와 이야기를 하며 `@RestController`라는 어노태이션에 `@ResponseBody`가 포함되어있고, 여기서 이미 반환데이터를 response body 에 담아서 준다는 의미가 포함되어있어서 response body에 정보를 담아준다는 의미가 중복된다는 의견을 들었습니다. 그리고 이 내용에 대해서 어느정도 동의가 된다고 생각을 해서 ResponseEntity를 사용하지 않고 HttpServletResponse를 통해서 상태 코드를 보내주도록 하였습니다.

혹시 RestController와 ResponseEntity를 같이 사용하는것이 위 의견처럼 의미의 중복이여서 피하는것이 좋은지, 아니면 그냥 사용해도 상관이 없을지에 대하여 의견을 들을 수 있을까요..?

ohjoohyung replied:
우선 ReponseEntity에 대한 좋은 [글](https://tecoble.techcourse.co.kr/post/2021-05-10-response-entity/)이 있으니 한번 참고 부탁드릴게요.
`@RestController`를 사용하게 되면 `@ResponseBody`에 의해 응답 값을 직렬화해 body에 담아 보낼 수 있게 되는데요. 물론 그대로 보내도 되지만 상태 코드나 header 등은 따로 설정을 해줘야 해요.  
ResponseEntity는 body 뿐만 아니라 header, 상태 코드 등을 손쉽게 응답 값으로 설정할 수 있도록 스프링에서 제공해주는 유용한 도구예요. 이걸 사용하지 않게 되면 필립이 작성해신 것처럼 header나 상태 코드를 설정하기 위해 HttpServletResponse를 직접 사용하거나 @ReponseStatus 등을 사용해 상태 코드를 적절하게 내려줘야 하는데 그것 보다는 ResponseEntity를 사용하는게 더 편리할 것 같습니다.

ohjoohyung commented on src/main/java/cart/controller/ProductsApiController.java:
```diff
@@ -0,0 +1,82 @@
+                .price(productRequestDto.getPrice())
+                .image(productRequestDto.getImage())
+                .build());
+        httpServletResponse.setStatus(HttpStatus.CREATED.value());
```
http 201 상태 코드 ([created](https://developer.mozilla.org/ko/docs/Web/HTTP/Status/201))에 대해 더 공부해보면 좋을 것 같네요!

pilyang replied:
단순히 생성되면 201로 보낸다 까지만 보고 이후 내용을 자세히 공부하지 못했었네요.. 😭 
`Location`헤더에 추가된 상품 위치를 `/products/{id}`값으로 추가해서 반환하도록 수정하였습니다.

ohjoohyung commented on src/main/java/cart/controller/dto/ProductRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.PositiveOrZero;
+
+public class ProductRequestDto {
```
dto 라고 꼭 dto를 붙이지 않아도 될 것 같아요.
http 요청, 응답에 필요한 dto니까 그냥 ~Request, ~Response로 네이밍을 가져가도 좋다고 생각합니다.

pilyang replied:
넵!! 네이밍 수정 했습니다!

ohjoohyung commented on src/main/java/cart/controller/dto/ProductRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+
+public class ProductRequestDto {
+
+    @NotNull(message = "name 필드가 있어야 합니다.")
+    private final String name;
+    @NotNull(message = "price 필드가 있어야 합니다.")
+    @PositiveOrZero(message = "price는 음수가 될 수 없습니다.")
+    private final Integer price;
```
```suggestion
    private final String name;
    
    @NotNull(message = "price 필드가 있어야 합니다.")
    @PositiveOrZero(message = "price는 음수가 될 수 없습니다.")
    private final Integer price;
    
```
개행 부탁드릴게요!

pilyang replied:
헉.. 수정 했습니다!

ohjoohyung commented on src/main/java/cart/dao/JdbcProductsDao.java:
```diff
@@ -0,0 +1,67 @@
+                        new Product(rs.getLong("id"),
+                                rs.getString("product_name"),
+                                rs.getInt("product_price"),
+                                rs.getString("product_image"))
+
+        );
```
```suggestion
                                rs.getString("product_image"))
        );
```

pilyang replied:
수정했습니다!

ohjoohyung commented on src/main/java/cart/dao/JdbcProductsDao.java:
```diff
@@ -0,0 +1,67 @@
+                product.getImage(),
+                product.getId()
+        );
+        if (updatedCount == 0) {
+            throw new NoSuchProductException("해당 상품이 없습니다.");
+        }
```
조금 더 객체지향적으로 접근해보면 좋을 것 같아요.
`db update 후 update한 row 수를 가져와 상품이 존재했는지를 판단하는 것`과 
`식별자 (id) 를 이용해 db에 있는 상품 가져오기 (없으면 예외) -> 가져온 상품 객체 update -> update 쿼리 실행
`
과는 무슨 차이가 있을까요?
지금 보니까 Product가 도메인인데 아무런 로직이 없더라고요ㅠ.ㅠ 필드 값을 수정해주는 기능이 있는데도요!
그리고 이렇게 서비스 로직을 둠으로써 db가 변경되었을 때 (ex db자체가 sql -> nosql로) 어떤 장점이 있을지도 생각해보면 좋겠네요.

참고로 이런 예외의 경우는 입력한 값도 예외 메시지에 같이 출력해주는게 더 좋습니다ㅎㅎ

pilyang replied:
우선 dao에 `findById`를 만들고 `ProductService`에서 findById를 통해 Product를 가져오고, 이때 여기서 존재하는 상품인지에 대한 확인이 이루어 지도록 수정 했습니다. 그리고 service의 update메서드에서 찾은 product를 dao의 update를 통해 변경하도록 수정했습니다.

기존에 했던 방식은 너무 db의 pk에 의존해서 1개의 정보만 바뀌거나 안바뀔것이라는 전제를 가지고 구현을 한 것 같습니다. 어떤식의 db를 사용하고 어떤 방식으로 접근하는지를 모르는 서비스 입장에서는 우선 변경할 상품의 아이디를 통해 상품을 가져오고 이때 상품이 없으면 예외를 발생, 있으면 가져온 상품의 정보 업데이트를 진행한다. 라는 방식이 맞았던 것 같습니다..!


첫 리뷰요청에서 보니 Product entity가 모든 필드를 final로 가지고 있었습니다... 그런데 update등 상품의 정보가 바뀌는 경우가 생기고, 이때 db의 정보와 Product객체의 정보를 동기화 시켜주어야 한다고 생각되어 id를 제외한 값에 final을 제거하였습니다!

추가로 각 값에 대한 검증은 처음 만들때는 어차피 클라이언트로 입력받은 값으로 상품을 만들고, 해당 내용들에 대한 검증은 requestBody에서 진행을 하였기에 Product에 잘못된 정보가 들어갈 일은 없겠지라는 막연한 생각에 검증로직을 빼게 되었습니다..

그런데 지금단계가 아닌 이후 비지니스 로직이 추가되고 하며 Product 객체를 사용하게 되면 개발자가 사용하며 문제가 생길 수 있어서 검증 로직이 필요한게 맞는 것 같습니다. 그래서 해당 도메인에 검증로직을 추가했습니다.

---

Product에 setter를 만들기 전 setter를 만들기와 정보가 변경된 새로운 Product객체를 만들어 반환하고, 내부 필드들은 final을 유지하기 중에서 고민을 했습니다.
고민 결과 setter를 만들게 된 이유는, 만약에 바뀐 정보를 가진 Product를 반환하도록 하면, 정보를 바꿔 새로운 상품 객체를 받더라도 기존의 Product객체는 같은 id를 가지지만 정보가 변하기 전 데이터를 가지고 있게 되어 이 객체의 사용시 문제가 생길 수 있다고 생각해 기존의 객체에 정보를 바꾸는 setter를 사용하게 되었습니다.  
이 선택에 대한 확신이 없어서,, 혹시 이에대한 의견 주실 수 있으시면 부타드리겠습니다.

ohjoohyung replied:
네네 기본적으로는 도메인에 대한 검증은 있어야 잘못된 도메인을 만들지 않을 수 있다고 생각해요.
사실 애플리케이션에서 제일 중요한 부분은 도메인이기도 하고요.
setter의 경우 무분별하게, 의미없이 사용하지 말라는거지 필요한 상황에서는 사용해도 괜찮다고 생각합니다.
지금처럼 도메인의 상태를 수정이 필요할 때도 있으니까요.
다만, setXXX 라는 네이밍보다는 조금 더 의미를 부여한 메서드 네이밍이 좋을 것 같네요. changeXXX, updateXXX와 같은 형태로요.

ohjoohyung commented on src/main/java/cart/service/dto/ProductDto.java:
```diff
@@ -0,0 +1,61 @@
+package cart.service.dto;
+
+public class ProductDto {
```
이 dto가 질문주신 부분 같은데 여기서 한번 얘기 나눠보면 좋을 것 같아요.
먼저 ProductRequestDto를 그대로 사용하지 않고 ProductDto를 만드신 이유를 알 수 있을까요?

pilyang replied:
dto의 경우 변경이 잦을 수 있고, 단순히 데이터를 전송한다는 의미만을 가지고 있어서 3개이상의 계층을 관통하게 되면 그 변화가 더 깊이까지 영향을 미치게 되어서 2계층간에만 이동할 수 있도록 하는것이 좋다고 알고 있습니다.
그런데 ProductRequetsDto의 경우는 외부 클라이언트와 서버 사이의 데이터 전송을 위한 dto임으로 해당 dto가 추가적으로 controller -> service등으로 더 깊이 들어가지 않게 하여 중간에 연결을 끊어주는것이 좋다고 생각을 하여 두 dto를 따로 두게 되었습니다.

ohjoohyung replied:
그렇군요. 그럼 서비스에서 요청, 응답 dto가 아닌 도메인을 받고 리턴하도록 만들고, 컨트롤러에서 dto -> 도메인으로 변환 후 넘겨줘도 되지 않을까요? 
제가 이렇게 말씀드리는 건 중간에 dto를 하나 더 두는 것이 과연 어떤 의미가 있는지에 대해 고민해보셨으면 싶어 여쭤봤습니다. 
지금 요구사항에 비해 너무 복잡한 설계는 아닐지에 대해서도 생각해보셨으면 하고요.

ohjoohyung commented on src/test/java/cart/controller/ProductsApiControllerTest.java:
```diff
@@ -0,0 +1,133 @@
+
+import static org.hamcrest.Matchers.is;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureTestDatabase
+class ProductsApiControllerTest {
+
+    private static final String NAME = "name";
+    private static final int PRICE = 0;
+    private static final String IMAGE = "https://i.namu.wiki/i/JP6VNIonP7Vbfs0ahYFE1NILgujcmmjemRN7lgHZEx1W7kFRcSvJ5buMh884QOP54Eg39LS7Ep-rizpE3Ny-obiZDGRHcy89Rtr4wEXzZmTW_qCORTIYmHmd9F8qavqkuXv-BvCR4zKkNwFm2MJ5LA.webp";
+
+    @LocalServerPort
+    private int port;
```
RestAssured 이용해서 인수테스트를 작성해주셨군요! 👍
인수테스트 클래스의 네이밍은 컨트롤러를 테스트하는게 아니다보니 ~AcceptanceTest로 가져가도 좋을 것 같네요!

pilyang replied:
감사합니다!! 클래스명 변경 했습니다! 변경하며 패키지도 `cart/controller` 에서 그냥 `cart`패키지로 변경 했습니다.

ohjoohyung commented on src/test/java/cart/controller/ProductsApiControllerTest.java:
```diff
@@ -0,0 +1,133 @@
+        RestAssured.given().log().all()
+                .contentType(ContentType.JSON)
+                .body(new ProductRequestDto("달리", PRICE, IMAGE))
+                .when().put("/products/1")
```
1을 그대로 작성하는 건 하드코딩이지 않을까요? 🤔
상품을 추가하고 가져온 id를 이용해 변경이나 삭제 테스트를 작성해주셨으면 해요.

pilyang replied:
헉.. 네! 추가시 id를 반환하고 이를 이용해서 테스트 하도록 수정했습니다!!

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+        jdbcTemplate.execute("truncate table products");
+        jdbcTemplate.execute("alter table products alter id restart with 1");
+
+        final String name1 = "test1";
+        final int price1 = 1000;
+        final String image1 = "testSource2";
+        final String name2 = "test2";
+        final int price2 = 2000;
+        final String image2 = "testSource2";
+        final String sql = "insert into products(product_name, product_price, product_image) values (?, ?, ?)";
+        jdbcTemplate.update(sql, name1, price1, image1);
+        jdbcTemplate.update(sql, name2, price2, image2);
```
모든 테스트에서 이 데이터가 필요하지 않는데 BeforeEach로 세팅할 필요가 있을까요?
필요한 곳에서 테스트의 given에 데이터를 세팅하는 것은 어떨까요?
그러면 테스트에서 필요한 데이터가 무엇인지 더 명확하게 보일 것 같습니다.

pilyang replied:
넵 필요한 부분에서만 설정해두는것이 좋을 것 같아서 해당 부분 변경했습니다.
다만, 데이터 초기화를 본 테스트클래스에서만 사용하지 않을 수 있다고 판단해 `testData.sql`파일을 만들어 초기화 쿼리를 두고, 필요한 부분에서 `@Sql`을 통해 초기화 하도록 했습니다.

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.NONE)
+@AutoConfigureTestDatabase
+class JdbcProductsDaoTest {
```
`@JdbcTest`에 대해서도 한번 찾아보면 좋을 것 같아요!

pilyang replied:
좋은 키워드 주셔서 감사합니다!!
찾아보니 `@JdbcTest`에서 제가 추가했던 `@AutoConfigureTestDatabase`를 비롯하여 db테스트를 위한 어노테이션들이 있음을 확인했습니다.
그리고 `@SpringBootTest`와 마찬가지로 application context를 가져오는데, 이때 spring boot test와 달리 `JdbcTestContextBootstrapper`를 통해 다른 컴포넌트들은 빈으로 생성하지 않고, jdbc테스트에 관련된 컴포넌트들만 가져오는것 같습니다.

다른 필요없는 빈들을 생성 안하고 테스트하고, 관련 설정들이 적용되어있어 jdbc관련 테스트에서는 해당 어노테이션을 사용하는것이 좋을 것 같아 `@JdbcTest`를 사용하도록 수정했습니다.

해당 부분을 변경해서 `JdbcProductDao`또한 불러오지 못해서 이를 빈으로 등록해주는 test configuration을 작성했습니다.
이런식으로 필요한것만 가져올 수 있도록 다시 configuration을 만드는것이 좋을지 아니면 `@BeforeEach`를 통해 `productDao`는 빈으로 관리하지 말고 매 테스트 전에 입력받은 주입받은 jdbcTemplate을 통해 매번 productDao에 `JdbcProductDao`를 생성해주는것이 좋았을지 의문입니다. 혹시 더 좋다고 생각하는 방법이 있으면, 아니면 이 방법들 말고 다른 방법이 있으면 조언해주시면 감사할 것 같습니다!!

ohjoohyung replied:
두 가지 방식 모두 사용해도 괜찮지만 지금은 크게 복잡하지 않으니 JdbcProductDao를 beforeEach에서 생성해주는 방식을 사용해도 괜찮다고 봅니다.

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+        jdbcTemplate.update(sql, name2, price2, image2);
+    }
+
+    @Test
```
`@DisplayName`으로 어떤 테스트인지 명시해주셨으면 합니다!

pilyang replied:
네 수정했습니다!!!

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+        final String image = "testSource";
+
+        // when
+        productsDao.create(name, price, image);
+
+        // then
+        final Map<String, Object> actual = jdbcTemplate.queryForMap("SELECT * FROM products ORDER BY id DESC LIMIT 1");
+        SoftAssertions.assertSoftly(softly -> {
+            softly.assertThat(actual.get("product_name")).isEqualTo(name);
+            softly.assertThat(actual.get("product_price")).isEqualTo(price);
+            softly.assertThat(actual.get("product_image")).isEqualTo(image);
```
create로 상품 id를 가져오고 이걸 이용해 조회하고 데이터를 검증해보는 것은 어떨까요?

pilyang replied:
POST에서 creat 201에 대한 리뷰를 반영하며 데이터 추가시 생성된 아이디를 반환하도록 리팩터링 했었고, 그렇게 반환되는 id를 이용하여 테스트를 할 수 있도록 고쳤습니다!

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,101 @@
+        final String newImage = "newSource";
+        final Product updateProduct = new Product(id, newName, newPrice, newImage);
+
+        final List<Long> selectIdFromProducts = jdbcTemplate.queryForList("SELECT id FROM products", Long.class);
+        System.out.println(selectIdFromProducts);
```
🤔

pilyang replied:
😭

ohjoohyung reviewed (changes requested):
안녕하세요 필립
장바구니 미션을 같이 하게 된 오즈라고 합니다.
1단계 기능 잘 구현해주셨어요! 👍
질문 답변 및 몇 가지 코멘트 남겼으니 확인 부탁드릴게요.
궁금한 점은 편하게 DM 보내주셔도 괜찮습니다.
수고하셨습니다~

pilyang committed:
refactor: image -> imageUrl 필드 명 변경

pilyang committed:
refactor: @RestControllerAdvice 적용

pilyang committed:
feat: 상품 추가 POST요청시 response header에 생성된 상품의 Location 추가

pilyang committed:
refactor: RequestBody, ResponseBody dto 클래스 명 변경

pilyang committed:
refactor: product update시 아이디 검증 로직 수정

pilyang committed:
feat: id로 상품 검색 기능 추가, 상품 update 로직 수

- ProductService update 로직 변경
- JdbcProductsDao
 - 아이디로 상품 검색 기능 추가
 - 업데이트, 상품 존재 확인 방식 수정
- 예외 메시지 확인 테스트 오류 수정

pilyang committed:
refactor: 상품 검증 추가

- Product 검증 추가
- ProductRequest 이름 검증 조건 변경

pilyang committed:
refactor: test class 명, 패키지 변경

pilyang committed:
fix: 테스트 이미지 주소 형식에 맞게 변경

pilyang committed:
test: JdbcProductsDaoTest 테스트 추가 및 테스트 데이터 추가 방식 변경

- findById 테스트 추가
- 데이터 초기화 방식 및 테스트 어노테이션 변경

pilyang committed:
test: api 테스트시 상품 id로 테스트하도록 변경

pilyang reviewed (commented):
안녕하세요 오즈!!

리뷰 남겨주신 내용들 보면서 놓친부분들을 많이 알 수 있었고 찾아볼 수 있었습니다.

리뷰 반영하면서 수정한 내용들과, 수정하며 궁금했던 내용들은 코멘트에 같이 남겨두었습니다.

감사합니다!!

ohjoohyung commented on src/main/java/cart/entity/Product.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    public void setName(final String name) {
+        validateName(name);
+        this.name = name;
+    }
+
+    public void setPrice(final int price) {
+        validatePrice(price);
+        this.price = price;
+    }
+
+    public void setImageUrl(final String imageUrl) {
+        validateImageUrl(imageUrl);
+        this.imageUrl = imageUrl;
+    }
```
setter에서 validation을 추가하셨는데 이렇게 되면 필드가 늘어나고 수정할 일이 생기면 매번 추가해야하지 않을까요?
이것보다는 현재는 Product를 수정할 때 전체 입력 값을 받아서 수정하니 update(Product product) 이런 형태로 두면 어떨까요? 
애초에 유효성 검증을 거친 Product만 인자로 들어와 수정에 사용될 수 있으니 검증 로직을 제거할 수 있을 것 같아요.

ohjoohyung commented on src/test/java/cart/dao/JdbcProductsDaoTest.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Test
+    @DisplayName("id로 검색 성공 테스트")
+    @Sql("classpath:testData.sql")
```
아 제 의도는 beforeEach에서 데이터를 세팅(create)해주는 부분을 각 테스트에 given으로 옮기는게 어떨까 였어요! 😂
전체 테스트에서 데이터 초기화는 필요하니 truncate, alter는 beforeEach에 그대로 두고요.

ohjoohyung reviewed (approved):
안녕하세요 필립
피드백 잘 반영해주셨어요! 👍
몇 가지 코멘트 남겼으니 확인 부탁드릴게요.
1단계는 충분하다고 생각하여 머지하고 2단계에서 뵙겠습니다.
수고하셨습니다~
